app/src/main/java/com/github/sdp/ratemyepfl/database/reviewable/ReviewableRepositoryImpl.kt

```diff
@@ -0,0 +1,118 @@
+) : ReviewableRepository<T>, LoaderRepository<T> by repository {
+
+    constructor(repository: Repository<T>, idFieldName: String) : this(
+        LoaderRepositoryImpl(repository),
+        idFieldName
+    )
+
+    constructor(
+        database: FirebaseFirestore,
+        collectionPath: String,
+        idFieldName: String,
+        transform: (DocumentSnapshot) -> T?
+    ) : this(
+        LoaderRepositoryImpl(RepositoryImpl(database, collectionPath, transform)),
+        idFieldName
+    )
+
+    companion object {
+        const val NUM_REVIEWS_FIELD_NAME = "numReviews"
+        const val AVERAGE_GRADE_FIELD_NAME = "averageGrade"
+
+        /**
+         * Compute the updated rating of the provided item if we add the provided rating.
+         *
+         *  @param item: the item for which we compute
+         *  @param rating: rating to add the the item
+         *
+         *  @return a [Pair] containing the computed number of reviews and average grade.
+         */
+        inline fun <reified T : Reviewable> computeUpdatedRating(
+            item: T,
+            rating: ReviewRating
+        ): Pair<Int, Double> {
+            val numReviews = item.numReviews
+            val averageGrade = item.averageGrade
+            val newNumReviews = numReviews + 1
+            val newAverageGrade: Double =
+                averageGrade + (rating.toValue() - averageGrade) / newNumReviews
+
+            return Pair(newNumReviews, newAverageGrade)
+        }
+    }
+
+
+    private val queryMostRated = query()
+        .orderBy(NUM_REVIEWS_FIELD_NAME, Query.Direction.DESCENDING)
+        .orderBy(idFieldName)
+
+    private val queryBestRated = repository
+        .query()
+        .orderBy(AVERAGE_GRADE_FIELD_NAME, Query.Direction.DESCENDING)
+        .orderBy(idFieldName)
+
+    /**
+     * Load a given number of [Reviewable] by decreasing number of reviews.
+     *
+     * @param number: number of item to load
+     *
+     * @return a [QueryResult] containing the result as a list of reviewable
+     */
+    override fun loadMostRated(number: UInt): QueryResult<List<T>> = load(
+        query()
+            .orderBy(NUM_REVIEWS_FIELD_NAME, Query.Direction.DESCENDING)
+            .orderBy(idFieldName), number
+    )
+
+    /**
+     * Load a given number of [Reviewable] by decreasing average grade.
+     *
+     * @param number: number of item to load
+     *
+     * @return a [QueryResult] containing the result as a list of reviewable
+     */
+    override fun loadBestRated(number: UInt): QueryResult<List<T>> = load(queryBestRated, number)
+
+
+    /**
+     * Search for a matching prefix in a provided field.
+     *
+     * @param prefix: the prefix to match
+     * @param field: the field where the match occurs
+     *
+     * @return a [QueryResult] containing a [List] of matched values. It matches at most
+     * [LIMIT_QUERY_SEARCH] values.
+     */
+    fun search(field: String, prefix: String): QueryResult<List<T>> =
```

Comment:
Shouldn't this repo implement the Searchable interface ? I was expecting an  'override' here.

Reply:
The Searchable interface fixes the field where we search things (or a set of fields). It's rather a utility function for repository decorating this repo.